akka-http2-support/src/main/scala/akka/http/impl/engine/http2/PriorKnowledgeSwitch.scala

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.impl.engine.http2
+
+import javax.net.ssl.SSLException
+
+import akka.util.ByteString
+import akka.NotUsed
+import akka.annotation.InternalApi
+import akka.http.impl.engine.server.HttpAttributes
+import akka.http.scaladsl.model.{ HttpRequest, HttpResponse }
+import akka.stream.TLSProtocol.{ SessionBytes, SessionTruncated, SslTlsInbound, SslTlsOutbound }
+import akka.stream.scaladsl.Flow
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+import akka.stream._
+
+/** INTERNAL API */
+@InternalApi
+private[http] object PriorKnowledgeSwitch {
+  type HttpServerFlow = Flow[ByteString, ByteString, NotUsed]
+  type HttpServerShape = FlowShape[ByteString, ByteString]
+
+  private final val PRIOR_KNOWLEDGE_PREFACE = ByteString("PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n")
+
+  def apply(
+    http1Stack: HttpServerFlow,
+    http2Stack: HttpServerFlow): HttpServerFlow =
+    Flow.fromGraph(
+      new GraphStage[HttpServerShape] {
+
+        // --- outer ports ---
+        val netIn = Inlet[ByteString]("PriorKnowledgeSwitch.netIn")
+        val netOut = Outlet[ByteString]("PriorKnowledgeSwitch.netOut")
+        // --- end of outer ports ---
+
+        override val shape: HttpServerShape =
+          FlowShape(netIn, netOut)
+
+        override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+          logic =>
+
+          // --- inner ports, bound to actual server in install call ---
+          val serverDataIn = new SubSinkInlet[ByteString]("ServerImpl.netIn")
+          val serverDataOut = new SubSourceOutlet[ByteString]("ServerImpl.netOut")
+          // --- end of inner ports ---
+
+          override def preStart(): Unit = pull(netIn)
+
+          setHandler(netIn, new InHandler {
+            private[this] var grabbed = ByteString.empty
+            def onPush(): Unit = {
+              val data = grabbed ++ grab(netIn)
+              if (data.length >= PRIOR_KNOWLEDGE_PREFACE.length) { // We should know by now
+                if (data.startsWith(PRIOR_KNOWLEDGE_PREFACE, 0))
+                  install(http2Stack, data)
+                else
+                  install(http1Stack, data)
+              } else if (data.isEmpty || data.startsWith(PRIOR_KNOWLEDGE_PREFACE, 0)) { // Still unknown
```

Comment:
`data.startsWith(PRIOR_KNOWLEDGE_PREFACE, 0)` will always be false here, since `data.length >= PRIOR_KNOWLEDGE_PREFACE.length` was already `false`, I think?

Reply:
(looks like the intent was to check whether the data received so far matches the preface, so we can switch to http1 as soon as we see a mismatch - which makes sense)

Reply:
@raboof Indeed!